File: src/test/java/com/hileco/cortex/analysis/processors/ProcessorFuzzTest.java
package com.hileco.cortex.analysis.processors;

import com.hileco.cortex.analysis.GraphBuilder;
import com.hileco.cortex.vm.VirtualMachine;
import com.hileco.cortex.vm.Program;
import com.hileco.cortex.vm.ProgramContext;
import com.hileco.cortex.vm.layer.LayeredMap;
import com.hileco.cortex.fuzzer.ProgramGenerator;
import com.hileco.cortex.instructions.ProgramException;
import com.hileco.cortex.instructions.ProgramRunner;
import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class ProcessorFuzzTest {

    private static final int LIMIT_RUNS = 500;

    private ProgramContext executeAll(LayeredMap<BigInteger, Program> atlas) {
        var caller = new Program(BigInteger.ZERO, new ArrayList<>());
        var callerContext = new ProgramContext(caller);
        for (var programAddress : atlas.keySet()) {
            var program = atlas.get(programAddress);
            var programContext = new ProgramContext(program);
            var processContext = new VirtualMachine(callerContext, programContext);
            try {
                var programRunner = new ProgramRunner(processContext);
                programRunner.run();
            } catch (ProgramException ignored) {
            }
        }
        return callerContext;
    }

    /**
     * Fuzzes the processor and verifies correctness by executing non-processed and processed versions
     * of generated programs, comparing for any differences in call results, storage, or transfers.
     */
    private void fuzzTestProcessor(Processor processor) {
        var graphBuilder = new GraphBuilder(List.of(
                new ParameterProcessor(),
                new FlowProcessor(),
                processor
        ));

        var seed = System.currentTimeMillis() * LIMIT_RUNS;
        long runs = 0;
        while (runs++ < LIMIT_RUNS) {
            var programGenerator = new ProgramGenerator();
            var runSeed = seed + runs;
            var generated = programGenerator.generate(runSeed);
            var generatedOptimized = programGenerator.generate(runSeed);
            for (var address : generatedOptimized.keySet()) {
                var program = generatedOptimized.get(address);
                var instructions = program.getInstructions();
                var optimizedInstructions = graphBuilder.build(instructions).toInstructions();
                generatedOptimized.put(program.getAddress(), new Program(program.getAddress(), optimizedInstructions));
            }
            var callerContext = this.executeAll(generated);
            var callerContextOptimized = this.executeAll(generatedOptimized);
            Assert.assertEquals(String.format("Issue with runSeed %d in caller", runSeed), callerContext.getMemory(),
                                callerContextOptimized.getMemory());
            Assert.assertEquals(String.format("Issue with runSeed %d in caller", runSeed), callerContext.getProgram().getStorage(),
                                callerContextOptimized.getProgram().getStorage());
            Assert.assertEquals(String.format("Issue with runSeed %d in caller", runSeed), callerContext.getProgram().getTransfers(),
                                callerContextOptimized.getProgram().getTransfers());
            for (var address : generated.keySet()) {
                var standard = generated.get(address);
                var optimized = generatedOptimized.get(address);
                Assert.assertEquals(String.format("Issue with runSeed %d in program %s", runSeed, address.toString()),
                                    standard.getStorage(),
                                    optimized.getStorage());
                Assert.assertEquals(String.format("Issue with runSeed %d in program %s", runSeed, address.toString()),
                                    standard.getTransfers(),
                                    optimized.getTransfers());
            }
        }
    }

    @Test
    public void fuzzTestExitTrimProcessor() {
        this.fuzzTestProcessor(new ExitTrimProcessor());
    }

    @Test
    public void fuzzTestJumpIllegalProcessor() {
        this.fuzzTestProcessor(new JumpIllegalProcessor());
    }

    @Test
    public void fuzzTestJumpThreadingProcessor() {
        this.fuzzTestProcessor(new JumpThreadingProcessor());
    }

    @Test
    public void fuzzTestJumpUnreachableProcessor() {
        this.fuzzTestProcessor(new JumpUnreachableProcessor());
    }

    @Test
    public void fuzzTestKnownJumpIfProcessor() {
        this.fuzzTestProcessor(new KnownJumpIfProcessor());
    }

    @Test
    public void fuzzTestKnownLoadProcessor() {
        this.fuzzTestProcessor(new KnownLoadProcessor(new HashMap<>(), new HashSet<>()));
    }

    @Test
    public void fuzzTestKnownProcessor() {
        this.fuzzTestProcessor(new KnownProcessor());
    }
}

File: src/test/java/com/hileco/cortex/analysis/processors/ProcessorFuzzTest.kt
package com.hileco.cortex.analysis.processors

import com.hileco.cortex.analysis.GraphBuilder
import com.hileco.cortex.fuzzer.ProgramGenerator
import com.hileco.cortex.instructions.ProgramException
import com.hileco.cortex.vm.concrete.Program
import com.hileco.cortex.vm.concrete.ProgramContext
import com.hileco.cortex.vm.concrete.ProgramRunner
import com.hileco.cortex.vm.concrete.VirtualMachine
import com.hileco.cortex.vm.layer.LayeredMap
import org.junit.Assert
import org.junit.Test
import java.math.BigInteger

abstract class ProcessorFuzzTest {
    private fun executeAll(atlas: LayeredMap<BigInteger, Program>): ProgramContext {
        val caller = Program(listOf())
        val callerContext = ProgramContext(caller)
        for (programAddress in atlas.keySet()) {
            val program = atlas[programAddress]
            val programContext = ProgramContext(program!!)
            val virtualMachine = VirtualMachine(callerContext, programContext)
            try {
                val programRunner = ProgramRunner(virtualMachine)
                programRunner.run()
            } catch (ignored: ProgramException) {
            }

        }
        return callerContext
    }

    /**
     * Fuzzes the processor and verifies correctness by executing non-processed and processed versions
     * of generated programs, comparing for any differences in call results, storage, or transfers.
     */
    private fun fuzzTestProcessor(processor: Processor) {
        val graphBuilder = GraphBuilder(listOf(
                ParameterProcessor(),
                FlowProcessor(),
                processor
        ))

        val seed = System.currentTimeMillis() * LIMIT_RUNS
        var runs: Long = 0
        while (runs++ < LIMIT_RUNS) {
            val programGenerator = ProgramGenerator()
            val runSeed = seed + runs
            val generated = programGenerator.generate(runSeed)
            val generatedOptimized = programGenerator.generate(runSeed)
            for (address in generatedOptimized.keySet()) {
                val program = generatedOptimized[address]
                val instructions = program!!.instructions
                val optimizedInstructions = graphBuilder.build(instructions).toInstructions()
                generatedOptimized[program.address] = Program(optimizedInstructions, program.address)
            }
            val callerContext = executeAll(generated)
            val callerContextOptimized = executeAll(generatedOptimized)
            Assert.assertEquals("Issue with runSeed $runSeed in caller", callerContext.memory, callerContextOptimized.memory)
            Assert.assertEquals("Issue with runSeed $runSeed in caller", callerContext.program.storage, callerContextOptimized.program.storage)
            Assert.assertEquals("Issue with runSeed $runSeed in caller", callerContext.program.transfers, callerContextOptimized.program.transfers)
            for (address in generated.keySet()) {
                val standard = generated[address]
                val optimized = generatedOptimized[address]
                Assert.assertEquals("Issue with runSeed $runSeed in program $address", standard!!.storage, optimized!!.storage)
                Assert.assertEquals("Issue with runSeed $runSeed in program $address", standard.transfers, optimized.transfers)
            }
        }
    }

    abstract fun fuzzTestableProcessor(): Processor

    @Test
    fun fuzzTest() {
        fuzzTestProcessor(fuzzTestableProcessor())
    }

    companion object {
        private const val LIMIT_RUNS = 500
    }
}
